### PR TITLE
Added RWLock on committed state

### PR DESCRIPTION
**Pull request by Shubham8287**
# Description of Changes
- Removed mutex Locking on wrapper state object (`Inner`) and added more granual locks on indivisual states.
- Introduced of RWlock on `CommittedState`.
- Moved implemented methods from `Inner` to `MutTxId` struct.
- Made bootstrap only dependent on CommittedState.

# API and ABI breaking changes

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

**Review by gefjon — CHANGES_REQUESTED** [review]
I gave up commenting on the `unwrap`s, but I'd like you to search through this change for every `unwrap` call, and:
- If you believe it never panics based only on local reasoning, add a comment describing why.
- If the caller must uphold some invariant to avoid a panic, add them as a doc comment to the function, and add a comment to the unwrap describing why those invariants are sufficient to avoid a panic.

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 1]
```diff
@@ -1538,33 +1715,65 @@ impl Inner {
...
+/// their own lock. To avoid deadlocks, it is crucial to acquire these locks
+/// in a consistent order throughout the application.
+///
+/// Lock Acquisition Order:
+/// 1. `memory`
+/// 2. `committed_state`
+/// 3. `tx_state`
+/// 4. `sequence_state`
+///
+/// All locking mechanisms are encapsulated within the struct through local methods.
```
Comment (gefjon) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:1740]: Love this!

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 2]
```diff
@@ -1574,29 +1800,28 @@ impl Locking {
...
         // Now we have to build our in memory structures.
-        datastore.build_sequence_state()?;
-        datastore.build_indexes()?;
+        commit_state.build_sequence_state(&mut sequence_state)?;
+        commit_state.build_indexes()?;
 
         log::trace!("DATABASE:BOOTSTRAPPING SYSTEM TABLES DONE");
-
+        drop(sequence_state);
+        drop(commit_state);
```
[the remark this answers is not in the document]
Comment (gefjon) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs:1805]: I wouldn't expect that to be the case - Rust is usually able to infer when borrows can/should/must end and insert drops automatically as appropriate.

**Comment by Shubham8287** [comment]
> I gave up commenting on the `unwrap`s, but I'd like you to search through this change for every `unwrap` call, and:
> 
> * If you believe it never panics based only on local reasoning, add a comment describing why.
> * If the caller must uphold some invariant to avoid a panic, add them as a doc comment to the function, and add a comment to the unwrap describing why those invariants are sufficient to avoid a panic.
 @gefjon  `unwrap()` has been used only for `tx_state` in `MutTxId`,  and `tx_state` must  not be None during whole transaction journey. So, It will never going to panic.

I feel it will redundant to add comment for it on every method but didn't found any good way to state this as well in comment, I added an comment for it https://github.com/clockworklabs/SpacetimeDB/pull/599/files#diff-92219f25d92114287cc7ccb95ba7bae63f6172afd393de14df7a44968b7fa952R125 on `MuTxId`, Let me know if you have better Idea :)

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 3]
```diff
@@ -115,8 +118,22 @@ impl<'a> DataRef<'a> {
     }
 }
 
+// Type aliases for lock gaurds
+type SharedWriteGuard<T> = ArcRwLockWriteGuard<RawRwLock, T>;
```
Comment (kulakowski) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:122]: I'll add that one reason I'm a fan of this is that it makes it much easier to change the concrete lock type in the future.

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs]: You are right but unexpectly for some reason, Compiler is not able to infer `sequence_state` (type `MutexGaurd`) last usage and expecting it to drop only when scope ends.
Though, I am able to get rid of explicit `drop` by not binding `sequence_state`, but looks something interesting happening here.

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs (hunk 3 above)
Comment (Shubham8287) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs:122]: Agree!

**Comment by Shubham8287** [comment]
> Are any updates to the threading model needed here?

@joshua-spacetime, this PR doesn't make any funcational changes, so No. May be my next part of "read write lock" would require them, Can you point me to code to understand bit about threading model.

**Review by gefjon — APPROVED** [review]
Great, thanks!

**Review by cloutiertyler — CHANGES_REQUESTED** [review]
It generally looks great. More or less exactly what I was looking for. I have a few comments inline. The most important one is that I think the Option can be removed around TxState now that it's no longer stored in the `Inner` struct.

Please see [this](https://github.com/clockworklabs/SpacetimeDB/pull/617) PR where I pulled it out. I'm 90% sure that this should work, but the 10% is because there's one place in `iter_by_col_range` that actually cared whether it was `None` or not. Could you confirm that my PR makes sense or that it doesn't?

**Comment by Shubham8287** [comment]
> I think the Option can be removed around TxState now that it's no longer stored in the `Inner` struct.

I had the same thoughts, which lead me to start discussion about dropping `transaction` object after `commit` or `rollback` to be sure about its lifecycle.

> Please see [this](https://github.com/clockworklabs/SpacetimeDB/pull/617) PR where I pulled it out. I'm 90% sure that this should work, but the 10% is because there's one place in `iter_by_col_range` that actually cared whether it was `None` or not. Could you confirm that my PR makes sense or that it doesn't?

`None` parts occurs only for bootstrapping, we have `CommittedState:chema_for_tablei` implemented for that, hence it should be fine.

let's move this conversaiton to your PR :)

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 4]
```diff
@@ -584,28 +450,307 @@ impl Inner {
         Ok(())
     }
 
+    // TODO(shubham): Need to confirm, if indexes exist during bootstrap to be used here.
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs:453]: My doubt is, If its possible to use Indexes during bootstrap we should do that, If we don't generate indexes by this time during bootstrap, we are fine as we are.  I just need to verify it, so added a TODO.

I have another though back of mind that we may be using this iter for read only transactions as well, then we have to add the indexes but that depends on the approach, little early to think about that here.

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 5]
```diff
@@ -584,28 +450,307 @@ impl Inner {
...
+    fn next(&mut self) -> Option<Self::Item> {
+        for (row_id, row) in self.iter.by_ref() {
+            let table_id = row.project_not_empty(self.table_id_col).unwrap();
+            if table_id == *self.value {
+                return Some(DataRef::new(row_id, row));
+            }
+        }
+        None
+    }
+}
```
[the remark this answers is not in the document]
Comment (Shubham8287) [reply crates/core/src/db/datastore/locking_tx_datastore/mod.rs:618]: I guess you mean to add endline before `None`

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 6]
```diff
@@ -669,7 +823,7 @@ impl Inner {
         Err(SequenceError::UnableToAllocate(seq_id).into())
     }
 
-    fn create_sequence(&mut self, seq: SequenceDef) -> super::Result<SequenceId> {
+    fn create_sequence(&mut self, seq: SequenceDef, database_address: Address) -> super::Result<SequenceId> {
```
Comment (kulakowski) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:826]: For next time: I think we could have factored this PR in two. Some of these changes are very mechanical, and IMO could have landed separately first. I'm always going to ask for more and smaller PRs.

**Review by cloutiertyler — APPROVED** [review]
This now LGTM.

**Review by mamcx — APPROVED** [review]
LGTM

**Pull request merged**